tests/function_objects/function_objects_fp16.cpp

```diff
@@ -0,0 +1,72 @@
+/*******************************************************************************
+//
+//  SYCL 2020 Conformance Test Suite
+//
+//  Copyright (c) 2023 The Khronos Group Inc.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+*******************************************************************************/
+
+#include "../../util/extensions.h"
+#include "../common/common.h"
+#include "../common/disabled_for_test_case.h"
+#include "../common/type_coverage.h"
+#include "function_objects.h"
+
+DISABLED_FOR_TEST_CASE(ComputeCpp, hipSYCL)
+("function objects void specializations scalar fp16",
+ "[function_objects][fp16]")({
+  auto queue = sycl_cts::util::get_cts_object::queue();
+  using availability = sycl_cts::util::extensions::availability<
+      sycl_cts::util::extensions::tag::fp16>;
+  if (!availability::check(queue)) {
+    SKIP("Device does not support half precision floating point operations.");
+  }
+
+  const auto types = named_type_pack<TYPES>::generate(TYPE_NAMES);
+  const auto types_only_half = named_type_pack<sycl::half>::generate("half");
+
+  // prevent testing duplicate combinations
+#if 0
```

Comment:
Why is this code disabled with preprocessor and not just removed?

Reply:
The code is disabled because of "Specification does not provide rules for the implicit conversion surrounding `sycl::half`. Since most types can be converted to `sycl::half` and vice versa, the type of the conditional operator is ambiguous." and not because of that comment, I have moved the comment so it's clear.

The reason it is disabled with the preprocessor is because the "setup" for the test is there, but the test itself cannot handle `half` because the specification is unclear about it. So, the structure of this file is correct but the test cannot be compiled for that data type. Once this specification problem is resolved, the `check_scalar_return_type` can be modified to handle `half` and the statements of this file can be re-enabled.

Reply:
This very minor, but what I'm trying to say is https://github.com/KhronosGroup/SYCL-Docs/issues/350 must be resolved before we can write a test. By the time https://github.com/KhronosGroup/SYCL-Docs/issues/350 is addressed, you might want to write a different test. 
Usually committing commented out code is considered as anti-pattern (e.g. https://kentcdodds.com/blog/please-dont-commit-commented-out-code).

Reply:
You are right, the test will have to be modified regardless of what is decided in 350. I have removed the half tests for function objects and added a comment in 350 stating that the test should be added once a decision is made.